doc/developer/design/20210629_source_trait.md

```diff
@@ -0,0 +1,319 @@
+    /// For example CREATE SOURCE FROM <name> HOST 'example.com'` would
+    /// correspond to `Sql` having a field `host` of type `String`.
+    ///
+    /// Types that implement `Deserialize` automatically implement `Deserialize`
+    type Sql: for<'de> Deserialize<'de>;
+
+    /// Type used to construct instances of this source
+    type Config: InstanceConfig;
+
+    /// The name of the source
+    fn name(&self) -> &'static str;
+
+    /// Constructs a new source from the parsed SQL. This is an opportunity for the implementation
+    /// to do any sanity check of the parameters passed
+    fn from_sql(sql: Self::Sql) -> Result<Self, Error>;
+
+    /// An opportunity for the source to remove any impurities from its
+    /// definition. This will be called before permanently storing the source in
+    /// the catalog. The default implementation is a no-op
+    async fn purify(&mut self) { }
+
+    /// Constructs a new source from the parsed SQL and WITH option. This is an
+    /// opportunity for the implementation to do any sanity check of the
+    /// parameters passed
+    fn new(sql_options: Self:Options, with_options: Self::WithOptions) -> Result<Self, Error>;
+
+    /// Creates at most `workers` instance configs. These configs will be
+    /// persisted in the catalog and be used to do any associated cleanup later.
+    /// It is up to the system to decide how the returned configs will be
+    /// distributed among the dataflow workers and converted into instances
+    fn create_instances(&self, workers: usize) -> Vec<Self::Config>;
+
+    /// The schema that will be produced by this source. The `RelationDesc`
+    /// returned conveys information of the name, type, and number of columns
+    /// but also of the key, if any exists.
+    fn desc(&self) -> RelationDesc;
+
+    /// If true, the source will only be available behind the experimental flag
+    fn is_experimental(&self) -> bool {
+        false
+    }
+
+    /// If false, the source will not be available in the cloud product
+    fn is_safe(&self) -> bool {
+        true
+    }
+}
+
+// `TimelyEvent` refers to `timely::dataflow::operators::to_stream::Event`
+pub type Event = TimelyEvent<Option<Timestamp>, Result<(Row, Timestamp, Diff), Error>>;
+
+#[async_trait]
+pub trait InstanceConfig: Serialize + for<'de> Deserialize<'de> {
+    /// Shared state between instances can be stored in this type for sources
+    /// that can share resources (e.g TCP sockets). A mutable reference to an
+    /// instance of `State` is passed to `Self::into_instance()`
+    type State: Default;
+
+    /// The instance type that will produce data into the dataflow from a ///
+    /// particular worker
+    type Instance: futures::Stream<Item = Event> + Send;
+
+    /// Transforms this configuration into an instance, re-using some of the
+    /// provided state if necessary
+    fn into_instance(self, state: &mut Self::State) -> Self::Instance;
```

Comment:
This is a moment where we will (I believe) want to add various `MapFilterProject` and `as_of` arguments, so that source instances can optionally absorb some of this implementation. We may also want the same functionality elsewhere, higher up, but maybe the corresponding arguments get jammed into the SQL arguments; I'm not sure (e.g., `CREATE SOURCE AS OF <TIME>`).